### PR TITLE
Add JSON serialize functionality

### PR DESCRIPTION
**Pull request by edominia**
Signed-off-by: Emilia Dominiak <emilia.dominiak@intel.com>

### CODE REVIEW

File: gef-integration.target/gef-integration.target [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
+<?pde version="3.8"?>
+<target name="gef-integration" sequenceNumber="1">
 <locations>
   <location includeAllPlatforms="false" includeConfigurePhase="true" includeMode="planner" includeSource="true" type="InstallableUnit">
     <unit id="org.eclipse.xtext.sdk.feature.group" version="0.0.0"/>
     <unit id="org.eclipse.fx.runtime.min.feature.feature.group" version="0.0.0"/>
     <unit id="org.eclipse.emf.sdk.feature.group" version="0.0.0"/>
     <unit id="org.eclipse.sdk.ide" version="0.0.0"/>
-    <repository location="http://download.eclipse.org/releases/2018-09"/>
+    <repository location="http://download.eclipse.org/releases/2019-09"/>
```
Comment (mmaka1) [inline gef-integration.target/gef-integration.target]: Pls do the changes related to version of local tools locally. Let's keep this definition with the base version until good to go with.

File: org.sofproject.alsa.topo/META-INF/MANIFEST.MF [hunk 2]
```diff
@@ -10,3 +10,4 @@ Export-Package: org.sofproject.alsa.topo.conf,
 Require-Bundle: org.sofproject.core,
  org.sofproject.topo.ui
 Bundle-Vendor: Intel Corporation
+Import-Package: org.sofproject.gst.json
```
Comment (mmaka1) [inline org.sofproject.alsa.topo/META-INF/MANIFEST.MF]: alsa.topo and gst are independent plug-ins, potentially used separately. There should be no dependencies introduced between them.

If you want to make it common for all topology plug-ins, and serialize through public interface of topology elements, it should be implemented inside topo.ui plug-in and communicate with gst.topo/alsa.topo like other elements of topo.ui do.

File: org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
+package org.sofproject.gst.json;
+import java.awt.*;
```
Comment (mmaka1) [inline org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java]: Do not use awt/swing. GUI should be based on swt as the other parts of Eclipse are.

File: org.sofproject.gst.topo/src/org/sofproject/gst/topo/model/GstTopoNode.java [hunk 4]
```diff
@@ -100,8 +100,8 @@ public GstTopoConnection getOutgoingConnection() {
 	public void serialize(Writer writer) throws IOException {
 		writer.write(elem.getName());
 		for (GstProperty prop : properties.values()) {
-			if (!prop.isChanged())
-				continue;
+			//if (!prop.isChanged())
```
Comment (mmaka1) [inline org.sofproject.gst.topo/src/org/sofproject/gst/topo/model/GstTopoNode.java]: Do not leave the dead code unless commented why it is kept here (e.g. some TODO).  (There might be existing bad examples in the code that require cleanup).

**Review by mmaka1 — CHANGES_REQUESTED** [review]

File: org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+import java.awt.event.*;
+import java.beans.*;
```
Comment (jajanusz) [inline org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java]: I'd not use wildcard imports at all

Comment (paulstelian97) [reply org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java]: IDE quirks maybe?

Comment (jajanusz) [reply org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java]: AFAIK IntelliJ and Eclipse will ask which one you want or default to no wildcards.

File: org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        setTitle("Serialize Json");
+
+		JLabel nameLabel = new JLabel("Name:");
+        nameField = new JTextField("");
+        
+		JLabel versionLabel = new JLabel("Version:");
+        versionField = new JTextField("");
+        
+		JLabel descriptionLabel = new JLabel("Description:");
+        descriptionField = new JTextField("");
```
Comment (jajanusz) [inline org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java]: Extra tabs

File: org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        
+        JLabel typeLabel = new JLabel("Type:");
+        String[] items = {"Gstreamer", "Ffmpeg"};
+        typeBox = new JComboBox<>(items);
+        
+        Object[] array = {nameLabel, nameField, versionLabel, versionField, descriptionLabel, descriptionField, typeLabel, typeBox};
+
+        Object[] options = {btnString1, btnString2};
```
Comment (jajanusz) [inline org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java]: formatting

File: org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    /**
+     * This method clears the dialog and hides it.
+     */
+    public void exit() {
+        dispose();
+    }
+
+}
```
Comment (jajanusz) [inline org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java]: Please add newline

File: org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java (hunk 5 above)
Comment (paulstelian97) [reply org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java]: @jajanusz I'm more versed in IntelliJ and I think it didn't ask me but it defaulted to doing wildcards. I know I have changed it to no wildcards when it did that and I have also done quick actions in the various Java projects I've done (college homework really). Not sure about Eclipse.

File: org.sofproject.gst.topo/src/org/sofproject/gst/topo/model/GstTopoGraph.java [hunk 9]
```diff
@@ -58,6 +62,9 @@
 import org.sofproject.topo.ui.graph.ITopoGraph;
 import org.sofproject.topo.ui.graph.ITopoNode;
 
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+
```
Comment (jajanusz) [inline org.sofproject.gst.topo/src/org/sofproject/gst/topo/model/GstTopoGraph.java]: Extra spaces

**Review by jajanusz — CHANGES_REQUESTED** [review]
Formatting issues - I'd go with default eclipse conventions as probably that will be what most of plugin developers will use.
So you probably can solve  of mentioned problems with `ctrl shif f` & `ctrl shift o` in eclipse.

File: org.sofproject.alsa.topo/src/org/sofproject/alsa/topo/model/AlsaTopoGraph.java [hunk 10]
```diff
@@ -536,7 +536,56 @@ public void serialize() throws CoreException, IOException {
...
 		outputFile.getParent().refreshLocal(1, null);
 	}
+	
+	@Override
+	public String getPipelineString() {
+		try {
+			ByteArrayOutputStream os = new ByteArrayOutputStream();
+			Writer writer = new BufferedWriter(new OutputStreamWriter(os));
+			// tlv-s
+			tlvs.serialize(writer);
```
Comment (jajanusz) [inline org.sofproject.alsa.topo/src/org/sofproject/alsa/topo/model/AlsaTopoGraph.java]: What value does `tlv-s` comment bring?

File: org.sofproject.alsa.topo/src/org/sofproject/alsa/topo/model/AlsaTopoGraph.java [hunk 11]
```diff
@@ -536,7 +536,56 @@ public void serialize() throws CoreException, IOException {
...
+
+			// vendor tokens
+			vTokensIndex.serialize(writer);
+
+			// vendor tuples
+			vTuplesIndex.serialize(writer);
```
Comment (jajanusz) [inline org.sofproject.alsa.topo/src/org/sofproject/alsa/topo/model/AlsaTopoGraph.java]: These comments are here just to decrypt variable name. Better just name them `vendorTokens/Tuples...` etc

File: org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+		typeCombo.setText(items[0]);
+		GridData typeGridData = new GridData(GridData.HORIZONTAL_ALIGN_FILL);
+		typeGridData.horizontalSpan = 3;
+		typeCombo.setLayoutData(typeGridData);
+
+		Button okButton = new Button(shell, SWT.PUSH);
+		okButton.setText("Ok");
+		GridData buttonGridData = new GridData(GridData.HORIZONTAL_ALIGN_FILL);
+		okButton.setLayoutData(buttonGridData);
+		okButton.addSelectionListener(new SelectionAdapter() {
```
Comment (jajanusz) [inline org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java:117]: I'm not a swt expert, but this stuff seems too verbose. Are there no more smart ways? Some builders/chaining at least? Also all these variables are introduced for 200 lines scope just to be used within ~5lines.

Comment (paulstelian97) [reply org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java:117]: @jajanusz I feel like this is generated code by some graphical tool. I'm not foreign to this kind of stuff even though I myself usually do that with C# rather than Java projects. The uniformity of actual code for each of the elements is a strong hint to me that this is generated code in fact so it doesn't deserve too deep of a review process honestly.

Comment (jajanusz) [reply org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java:117]: Usually IDEs give some marker comments in such case to recognize what's generated/manually added. But it guess different tools can have different ways of handling it.

File: gef-integration.target/gef-integration.target [hunk 13]
```diff
@@ -1,5 +1,6 @@
 <?xml version="1.0" encoding="UTF-8" standalone="no"?>
-<?pde version="3.8"?><target name="gef-integration" sequenceNumber="1">
+<?pde version="3.8"?>
+<target name="gef-integration" sequenceNumber="1">
```
Comment (jajanusz) [inline gef-integration.target/gef-integration.target]: Remove it or make it separate commit f.e. `gef-integration: fix .target formatting`

File: org.sofproject.gst.topo/.classpath [hunk 14]
```diff
@@ -1,7 +1,17 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <classpath>
-	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8">
+		<accessrules>
+			<accessrule kind="accessible" pattern="javafx/**"/>
+		</accessrules>
+	</classpathentry>
```
Comment (jajanusz) [inline org.sofproject.gst.topo/.classpath:7]: Why is javafx mentioned in 3 projects? You probably need it only in 1.

**Review by jajanusz — CHANGES_REQUESTED** [review]

File: org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java (hunk 12 above)
Comment (jajanusz) [reply org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java:117]: Can live without that change for now.

File: org.sofproject.alsa.topo/src/org/sofproject/alsa/topo/model/AlsaTopoGraph.java (hunk 10 above)
Comment (jajanusz) [reply org.sofproject.alsa.topo/src/org/sofproject/alsa/topo/model/AlsaTopoGraph.java]: Please refactor variable names or remove these comments, cos now they only waste readers time.

File: org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java (hunk 5 above)
Comment (jajanusz) [reply org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java]: @edominia thanks

**Comment by lgirdwood** [comment]
@edominia please also use a more descriptive commit message. You need to explain what, why and how for each patch.

File: org.sofproject.alsa.topo/src/org/sofproject/alsa/topo/model/AlsaTopoGraph.java [hunk 15]
```diff
@@ -490,37 +490,16 @@ public void serialize() throws CoreException, IOException {
...
-		// pcm-s
 		pcms.serialize(writer);
-
-		// be-s
 		beIndex.serialize(writer);
-
-		// hw-configs
 		hwConfigs.serialize(writer);
-
-		// pipelines (widgets + graphs)
```
Comment (jajanusz) [inline org.sofproject.alsa.topo/src/org/sofproject/alsa/topo/model/AlsaTopoGraph.java]: Looks like it was previously marked as your addition cos of spaces vs tabs issue. You can remove it now from this pr as it's not related.

File: org.sofproject.gst.topo/.classpath (hunk 14 above)
Comment (jajanusz) [reply org.sofproject.gst.topo/.classpath:7]: @edominia ping

**Comment by jajanusz** [comment]
Regarding jars that this PR adds f.e. - ` org.sofproject.gst.topo/lib/jackson-annotations-2.10.1.jar `
@mmaka1 Is it ok to have these binaries here, are there no licensing issues?

File: org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+		this.graph = graph;
+		Shell shell = new Shell(display, SWT.CLOSE | SWT.TITLE | SWT.MIN);
+
+		shell.setText("Serialize JSON");
+
+		GridLayout gridLayout = new GridLayout(4, false);
+		gridLayout.verticalSpacing = 8;
+		shell.setLayout(gridLayout);
+
+		new Label(shell, SWT.NULL).setText("Name:");
```
Comment (mmaka1) [inline org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java:83]: Name, description, and other properties collected here should be properties accessible via the ITopoGraph interface, displayed and editable in the Properties view. So that the serialize-to-json operation would not require any user interaction.

File: org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+package org.sofproject.gst.json;
```
Comment (mmaka1) [inline org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java:30]: If this is part of org.sofproject.gst.**topo** plugin, there should not be higher level packages inside.

File: org.sofproject.topo.ui/META-INF/MANIFEST.MF [hunk 18]
```diff
@@ -22,7 +22,8 @@ Import-Package: com.google.common.collect;version="21.0.0",
  com.google.inject;version="1.3.0",
  com.google.inject.binder;version="1.3.0",
  com.google.inject.multibindings;version="1.3.0",
- com.google.inject.util;version="1.3.0"
+ com.google.inject.util;version="1.3.0",
+ org.sofproject.gst.json
```
Comment (mmaka1) [inline org.sofproject.topo.ui/META-INF/MANIFEST.MF:26]: Do not create a dependency from a common base to a specific single topology plug-in. If json serialization is supposed to be a common functionality, add it to this topo.ui plugin.

File: org.sofproject.topo.ui/src/org/sofproject/topo/ui/editor/policies/TopoEditorOnClickHandler.java [hunk 19]
```diff
@@ -29,6 +29,7 @@
 
 package org.sofproject.topo.ui.editor.policies;
 
+import java.awt.Frame;
```
Comment (mmaka1) [inline org.sofproject.topo.ui/src/org/sofproject/topo/ui/editor/policies/TopoEditorOnClickHandler.java:32]: Not sure why would AWT be needed here?

File: org.sofproject.alsa.topo/src/org/sofproject/alsa/topo/model/AlsaTopoGraph.java [hunk 20]
```diff
@@ -536,7 +515,35 @@ public void serialize() throws CoreException, IOException {
 		}
 		outputFile.getParent().refreshLocal(1, null);
 	}
-
+	
+	@Override
+	public String getPipelineString() {
```
Comment (mmaka1) [inline org.sofproject.alsa.topo/src/org/sofproject/alsa/topo/model/AlsaTopoGraph.java]: It is a graph, that may consist of multiple pipelines, not a single one, so the method name should not refer to a pipeline.

File: org.sofproject.alsa.topo/src/org/sofproject/alsa/topo/model/AlsaTopoGraph.java [hunk 21]
```diff
@@ -536,7 +515,35 @@ public void serialize() throws CoreException, IOException {
...
+		try {
+			ByteArrayOutputStream os = new ByteArrayOutputStream();
```
Comment (mmaka1) [inline org.sofproject.alsa.topo/src/org/sofproject/alsa/topo/model/AlsaTopoGraph.java]: If you need a string, you may use the StringBuffer, not a byte representation that might be converted into characters and back.

File: org.sofproject.alsa.topo/src/org/sofproject/alsa/topo/model/AlsaTopoGraph.java [hunk 22]
```diff
@@ -536,7 +515,35 @@ public void serialize() throws CoreException, IOException {
...
+			Writer writer = new BufferedWriter(new OutputStreamWriter(os));
+			tlvs.serialize(writer);
```
Comment (mmaka1) [inline org.sofproject.alsa.topo/src/org/sofproject/alsa/topo/model/AlsaTopoGraph.java]: Do not duplicate the code of ITopoGraph.serialize(). Atm me may just refactor the latter to return a String and move target file content writing to topo.ui.

File: org.sofproject.gst.topo/build.properties [hunk 23]
```diff
@@ -2,4 +2,7 @@ source.. = src/
 output.. = bin/
 bin.includes = META-INF/,\
                .,\
-               plugin.xml
+               plugin.xml,\
+               lib/jackson-annotations-2.10.1.jar,\
```
Comment (mmaka1) [inline org.sofproject.gst.topo/build.properties:6]: Do not import external binaries to the plug-in repository. Add a dependency to an existing Eclipse plug-in(s) to the manifest (and adjust the documentation if any of them is not a part of core bundle,  see how the dependency on GEF is handled).

**Review by mmaka1 — CHANGES_REQUESTED** [review]

File: org.sofproject.gst.topo/.classpath (hunk 14 above)
Comment (edominia) [reply org.sofproject.gst.topo/.classpath:7]: It's mentioned only in two projects and it's necessary in both of them

**Comment by edominia** [comment]
> Regarding jars that this PR adds f.e. - `org.sofproject.gst.topo/lib/jackson-annotations-2.10.1.jar`
> @mmaka1 Is it ok to have these binaries here, are there no licensing issues?

Jackson is on Apache license so I think it's ok.

File: org.sofproject.alsa.topo/src/org/sofproject/alsa/topo/model/AlsaTopoGraph.java (hunk 20 above)
Comment (edominia) [reply org.sofproject.alsa.topo/src/org/sofproject/alsa/topo/model/AlsaTopoGraph.java]: ok

**Comment by edominia** [comment]
Merge three pull requests into one #43

File: org.sofproject.topo.ui/META-INF/MANIFEST.MF (hunk 18 above)
Comment (edominia) [reply org.sofproject.topo.ui/META-INF/MANIFEST.MF:26]: ok, done

File: org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java (hunk 17 above)
Comment (edominia) [reply org.sofproject.gst.topo/src/org/sofproject/gst/json/JsonCustomOptionPane.java:30]: ok, done
